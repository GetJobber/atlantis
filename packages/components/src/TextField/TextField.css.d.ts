export const small: string;
export const normal: string;
export const large: string;
export const error: string;
export const inputWrapper: string;
export const label: string;
export const input: string;
export const labelContent: string;
