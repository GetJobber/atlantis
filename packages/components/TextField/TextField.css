.TextField {
  position: relative;
  margin: 4px 0 1rem;
  font-size: 14px;
  line-height: 3;
  color: #012939;
}

.TextField::before {
  position: absolute;
  content: "";
  box-shadow: 0 0 0 2px #e0e0e0;
  transition: box-shadow 200ms cubic-bezier(0.075, 0.82, 0.165, 1);
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  border-radius: 4px;
  z-index: 0;
  pointer-events: none;
}

.small {
  line-height: 2;
}

.focused::before {
  box-shadow: 0 0 0 2px #ddc30f;
}

.large {
  font-size: 16px;
  line-height: 4;
}

.error::before {
  box-shadow: 0 0 0 2px #ef5733;
}

.multiline {
  padding-top: 1em;
  line-height: 1.5;
}
