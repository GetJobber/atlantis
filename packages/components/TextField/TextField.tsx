import classNames from "classnames";
import React, { InputHTMLAttributes, useState } from "react";
import FieldAffix from "./FieldAffix/FieldAffix";
import Input from "./Input/Input";
import Label from "./Label/Label";
import styles from "./TextField.css";

type Omit<T, K> = Pick<T, Exclude<keyof T, K>>;

type Size = "normal" | "small" | "large";

interface TextFieldProps
  extends Omit<
      InputHTMLAttributes<HTMLInputElement>,
      "size" | "defaultValue" | "prefix"
    > {
  /**
   * The text field label
   */
  label?: React.ReactNode | string;
  error?: boolean;
  disabled?: boolean;
  defaultValue?: string | number;
  size?: Size;

  multiline?: boolean;
  /**
   * Number of rows for a multiline text field
   */
  rows?: number;

  prefix?: React.ReactNode | string;
  suffix?: React.ReactNode | string;
}

export default function TextField(props: TextFieldProps) {
  const {
    children,
    className,
    name,
    label,
    required,
    defaultValue,
    multiline,
    size,
    error,
    prefix,
    suffix,
    ...other
  } = props;

  const { filled, ...handleInput } = useInputValue(defaultValue);
  const { focused, ...handleFocus } = useFocusState();

  const classes = classNames([
    styles[size],
    error && styles.error,
    focused && styles.focused,
    multiline && styles.multiline,
    styles.wrapper,
  ]);

  const showLabel: boolean = size !== "small" && label !== undefined;
  let labelElement;
  if (showLabel) {
    labelElement = (
      <Label htmlFor={name} error={error} shrink={filled}>
        {label}
      </Label>
    );
  }

  if (!filled) {
    other.placeholder = `${label}`;
  }

  return (
    <div className={classes}>
      <FieldAffix position="start">{prefix}</FieldAffix>
      <div className={styles.field}>
        {labelElement}
        <Input
          id={name}
          name={name}
          multiline={multiline}
          {...handleFocus}
          {...handleInput}
          {...other}
        />
      </div>
      <FieldAffix position="end">{suffix}</FieldAffix>
    </div>
  );
}

function useFocusState() {
  const [focused, setFocused] = useState(false);

  return {
    focused,
    onBlur: (): void => setFocused(false),
    onFocus: (): void => setFocused(true),
  };
}

function useInputValue(initialValue: string | number) {
  const [value, setValue] = useState(initialValue);
  const [filled, setFilled] = useState(initialValue && initialValue !== "");

  return {
    value,
    filled,
    onChange(e: React.ChangeEvent<HTMLInputElement>): void {
      setValue(e.target.value);
      setFilled(/\S/.test(e.target.value));
    },
  };
}

const defaultProps: TextFieldProps = {
  size: "normal",
  defaultValue: "",
  rows: 3,
};

TextField.defaultProps = defaultProps;
