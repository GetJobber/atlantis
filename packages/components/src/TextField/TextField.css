:root {
  --input-color: var(--color-blue);
  --label-text-color: var(--color-greyBlue--light);
  --input-border-width: 2px;
  --input-border-color: var(--color-grey--lighter);
  --input-background-color: var(--color-white);

  --input-border-color--focused: var(--color-yellow);

  --input-background-color--disabled: var(--color-grey--lightest);

  --label-text-color--error: var(--color-red);
  --input-border-color--error: var(--color-red);
}

.small {
  --input-height: var(--input-height--small, 36px);
}

.normal {
  --input-height: var(--input-height--normal, 48px);
}

.large {
  --input-height: var(--input-height--large, 64px);
  --input-font-size: var(--input-font-size--large, 1em);
}

.error {
  --input-border-color: var(--input-border-color--error);
  --input-border-color--focused: var(--input-border-color--error);
  --label-text-color: var(--label-text-color--error);
}

.inputWrapper {
  margin-bottom: 1em;
  font-family: var(
    --input-font-family,
    "Noto Sans",
    Helvetica,
    Arial,
    sans-serif
  );
  font-size: var(--root-font-size, 16px);
}

.inputWrapper:focus-within {
  --input-border-color: var(--input-border-color--focused);
}

.label,
.input::placeholder {
  color: var(--label-text-color);
}

.label {
  display: flex;
  position: relative;
  width: 100%;
  height: var(--input-height);
  font-size: var(--input-font-size, 0.875em);
}

.input {
  width: 100%;
  box-sizing: border-box;
  padding: 0 1em;
  border: var(--input-border-width) solid var(--input-border-color);
  border-radius: 4px;
  outline: none;
  color: var(--input-color);
  font-family: inherit;
  font-size: inherit;
  background-color: var(--input-background-color);
}

.input:disabled {
  --input-background-color: var(--input-background-color--disabled);
  opacity: 0.6;
}

.labelContent {
  position: absolute;
  left: 1em;
  width: 100%;
  padding: var(--input-border-width);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  pointer-events: none;
  opacity: 0;
  transform: none;
  transition: transform 100ms, opacity 100ms;
  align-self: center;
  transform-origin: top left;
}

.input:not(:placeholder-shown) + .labelContent {
  opacity: 1;
  transform: scale(0.7) translate(1px, -1em);
}
