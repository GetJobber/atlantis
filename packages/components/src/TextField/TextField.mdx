---
menu: Components
---

import { Playground, Props } from "docz";
import { TextField } from ".";

# Text Field

Text fields are used in forms to accept input from users.

## Basic Text Field

<Playground>
  <TextField size="small" placeholder="Small input" />
  <TextField placeholder="Normal input" />
  <TextField size="large" placeholder="Large input" />
</Playground>


## Text Field States

<Playground>
  <TextField disabled placeholder="Disabled input" />
  <TextField error placeholder="Invalid input" />
</Playground>

## Properties

<Props of={TextField} />
