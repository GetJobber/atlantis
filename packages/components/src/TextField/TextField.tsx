import React, { ChangeEventHandler } from "react";
import classnames from "classnames";
import styles from "./TextField.css";

interface TextFieldProps {
  /** The input name  */
  readonly name?: string;
  /** The text that appears when no value is set and displayed as a hover label when a value is present. */
  readonly placeholder?: string;
  /** The initial input value. */
  readonly defaultValue?: string;
  /** Callback fired when the value is changed. */
  readonly onChange?: ChangeEventHandler<HTMLInputElement>;
  /**
   * The size of the input
   * @default normal
   */
  readonly size?: "small" | "normal" | "large";
  /**
   * Indicates whether the input is disabled or not.
   * @default false
   */
  readonly disabled?: boolean;
  /**
   * Indicates whether the input is in an error state or not.
   * @default false
   */
  readonly error?: boolean;
}

export function TextField({
  name,
  placeholder,
  defaultValue,
  onChange,
  size = "normal",
  disabled = false,
  error = false,
}: TextFieldProps) {
  const wrapperClass = classnames(
    styles.inputWrapper,
    styles[size],
    error && styles.error,
  );

  return (
    <div className={wrapperClass}>
      <label className={styles.label} htmlFor={name}>
        <input
          type="text"
          className={styles.input}
          name={name}
          placeholder={placeholder}
          disabled={disabled}
          onChange={onChange}
          defaultValue={defaultValue}
        />
        {size !== "small" && placeholder && (
          <span className={styles.labelContent}>{placeholder}</span>
        )}
      </label>
    </div>
  );
}
