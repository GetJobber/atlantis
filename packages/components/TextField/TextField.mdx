---
name: TextField
menu: Components
route: /components/TextField
---

import { Playground, PropsTable } from "docz";
import TextField from "./TextField";
import Card from "@atlantis/icons/Card";

# Text Field

### Use the Text Field component to accept user input in forms.

## Basic Usage

The TextField component is a wrapper for an input and label with support for error states and adornments.

```
import TextField from "@atlantis/components/TextField";
```

<Playground>
  <TextField label="Placeholder field" />
</Playground>

## Default value

<Playground>
  <TextField label="Placeholder field" defaultValue="This is a default" />
</Playground>

## Sizing

<Playground>
  <TextField label="Small placeholder field" size="small" />
  <TextField label="Placeholder field" />
  <TextField label="Large placeholder field" size="large" />
  <TextField label="Small multiline placeholder field" size="small" multiline />
  <TextField label="Multiline placeholder field" multiline />
  <TextField label="Large multiline placeholder field" size="large" multiline />
</Playground>

## Affix

<Playground>
  <TextField label="Card Number" suffix={<Card />} />
  <TextField
    label="Multiline placeholder field"
    multiline
    prefix={<marquee>weeee</marquee>}
  />
</Playground>

## Errors

<Playground>
  <TextField error label="Oh boy, what have you done?" />
</Playground>

## Properties

<PropsTable of={TextField} />
