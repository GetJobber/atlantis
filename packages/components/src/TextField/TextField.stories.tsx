/* eslint-env node */
import * as React from "react";
import { storiesOf } from "@storybook/react";
import { TextField } from "./TextField";

const stories = storiesOf("Components", module);

stories.add("TextField", () => <TextField placeholder="Normal input" />, {
  info: { inline: true },
});
