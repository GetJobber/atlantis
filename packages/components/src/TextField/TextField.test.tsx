import React from "react";
import { TextField } from "./TextField";
import renderer from "react-test-renderer";

it("renders correctly with no props", () => {
  const tree = renderer.create(<TextField />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
    <div
      className="inputWrapper normal"
    >
      <label
        className="label"
      >
        <input
          className="input"
          disabled={false}
          type="text"
        />
      </label>
    </div>
  `);
});

it("renders correctly with a placeholder", () => {
  const tree = renderer
    .create(<TextField placeholder="My placeholder" />)
    .toJSON();
  expect(tree).toMatchInlineSnapshot(`
    <div
      className="inputWrapper normal"
    >
      <label
        className="label"
      >
        <input
          className="input"
          disabled={false}
          placeholder="My placeholder"
          type="text"
        />
        <span
          className="labelContent"
        >
          My placeholder
        </span>
      </label>
    </div>
  `);
});

it("renders correctly as small", () => {
  const tree = renderer.create(<TextField size="small" />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
    <div
      className="inputWrapper small"
    >
      <label
        className="label"
      >
        <input
          className="input"
          disabled={false}
          type="text"
        />
      </label>
    </div>
  `);
});

it("renders correctly as large", () => {
  const tree = renderer.create(<TextField size="large" />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
    <div
      className="inputWrapper large"
    >
      <label
        className="label"
      >
        <input
          className="input"
          disabled={false}
          type="text"
        />
      </label>
    </div>
  `);
});

it("renders correctly in a disabled state", () => {
  const tree = renderer.create(<TextField disabled />).toJSON();
  expect(tree).toMatchInlineSnapshot(`
    <div
      className="inputWrapper normal"
    >
      <label
        className="label"
      >
        <input
          className="input"
          disabled={true}
          type="text"
        />
      </label>
    </div>
  `);
});
