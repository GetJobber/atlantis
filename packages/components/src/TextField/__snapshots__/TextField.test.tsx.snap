// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="inputWrapper normal"
>
  <label
    className="label"
  >
    <input
      className="input"
      disabled={false}
      type="text"
    />
  </label>
</div>
`;
